File: ADT/djikstra's algo/shortest_path.cpp
// Shortest Path in Weighted undirected graph(GFG)
// Tc = O(m.logn)
// Sc = O(n + m)

class Solution {
public:
    vector<int> shortestPath(int n, int m, vector<vector<int>>& edges) {

        vector<vector<pair<int,int>>> adj(n+1);
        for (auto &vec : edges) {
            int u = vec[0], v = vec[1], w = vec[2];
            adj[u].push_back({v, w});
            adj[v].push_back({u, w});
        }

        priority_queue<pair<int,int>, vector<pair<int,int>>, greater<pair<int,int>>> pq;
        vector<int> dist(n+1, INT_MAX);
        vector<int> parent(n+1);

        for (int i = 1; i <= n; i++) parent[i] = i;

        dist[1] = 0;
        pq.push({0, 1});

        while (!pq.empty()) {
            auto [d, node] = pq.top();
            pq.pop();

            // Important: skip stale distances
            if (d != dist[node]) continue;

            for (auto &it : adj[node]) {
                int adjNode = it.first;
                int weight  = it.second;

                if (d + weight < dist[adjNode]) {
                    dist[adjNode] = d + weight;
                    parent[adjNode] = node;
                    pq.push({dist[adjNode], adjNode});
                }
            }
        }

        // No path exists
        if (dist[n] == INT_MAX) return {-1};

        // Reconstruct path
        vector<int> path;
        int node = n;

        while (parent[node] != node) {
            path.push_back(node);
            node = parent[node];
        }

        path.push_back(1);
        reverse(path.begin(), path.end());
        return path;
    }
};

File: ADT/djikstra's algo/shortest_path.py
#  Shortest Path in Weighted undirected graph(GFG)

import heapq

def shortestPath(n, m, edges):
    """
    Computes the shortest path from node 1 to node n using Dijkstra's algorithm.
    Returns the actual path (list of nodes) or [-1] if no path exists.
    """

    # ---- Build adjacency list ----
    adj = [[] for _ in range(n + 1)]
    for u, v, w in edges:
        adj[u].append((v, w))
        adj[v].append((u, w))

    # ---- Dijkstra algo ----
    dist = [float('inf')] * (n + 1)
    parent = list(range(n + 1))
    pq = []  # (distance, node)

    dist[1] = 0
    heapq.heappush(pq, (0, 1))

    # ---- Dijkstra's Algorithm ----
    while pq:
        d, node = heapq.heappop(pq)

        # Skip stale entries
        if d != dist[node]:
            continue

        for adjNode, weight in adj[node]:
            if d + weight < dist[adjNode]:
                dist[adjNode] = d + weight
                parent[adjNode] = node
                heapq.heappush(pq, (dist[adjNode], adjNode))

    # ---- No path to n ----
    if dist[n] == float('inf'):
        return [-1]

    # ---- Reconstruct the path ----
    path = []
    curr = n
    while parent[curr] != curr:
        path.append(curr)
        curr = parent[curr]
    path.append(1)
    path.reverse()

    return path
